src/modules/python/pbs_hooks/PBS_cray_jacs.PY

```diff
@@ -0,0 +1,339 @@
+    merged = {}
+    for key in base:
+        if key in newkeys and isinstance(base[key], dict):
+            # Take it off the list of keys to copy
+            newkeys.remove(key)
+            merged[key] = merge_dict(base[key], new[key])
+        else:
+            merged[key] = copy.deepcopy(base[key])
+    # Copy the remaining unique keys from new
+    for key in newkeys:
+        merged[key] = copy.deepcopy(new[key])
+    return merged
+
+
+def handle_execjob_begin():
+    """
+    Handler for execjob_begin events.
+    """
+    log_function_name()
+    event = HookHelper.get_event()
+    jid = event.job.id
+    uid = pwd.getpwnam(event.job.euser).pw_uid
+    log_with_caller(pbs.EVENT_DEBUG4, 'UID is %d' % uid)
+    data = {
+        'jobid': jid,
+        'uid': uid
+    }
+    url = HookHelper.build_path(resource='job')
+    timeout = HookHelper.get_config()['post_timeout']
+    try:
+        r = post(url, json=data, timeout=timeout)
+    except requests.Timeout:
+        log_with_caller(pbs.EVENT_ERROR, 'POST timed out for job %s' % jid)
+        raise RejectError()
+    if r.status_code != requests.codes['ok']:
+        log_with_caller(pbs.EVENT_ERROR, 'Invalid status code %d' %
+                        r.status_code)
+        raise RejectError()
+    log_with_caller(pbs.EVENT_DEBUG, 'Job %s registered' % jid)
+
+
+def handle_execjob_end():
+    """
+    Handler for execjob_end events.
+    """
+    log_function_name()
+    jid = HookHelper.get_event().job.id
+    url = HookHelper.build_path(resource='job', jobid=jid)
+    timeout = HookHelper.get_config()['delete_timeout']
+    try:
+        r = delete(url, timeout=timeout)
+        r.raise_for_status()
+    except requests.Timeout:
+        log_with_caller(pbs.EVENT_ERROR, 'POST timed out for job %s' % jid)
+        raise RejectError()
+    except requests.HTTPError:
+        log_with_caller(pbs.EVENT_ERROR, 'DELETE job failed')
+        raise RejectError('Job delete failed')
+
+    log_with_caller(pbs.EVENT_DEBUG, 'Job %s deleted' % jid)
+
+
+def main():
+    """
+    Main function for execution
+    """
+    log_function_name()
+    hostname = pbs.get_local_nodename()
+    # Log the hook event type
+    event = HookHelper.get_event()
+
+    handlers = {
+        pbs.EXECJOB_BEGIN: (handle_execjob_begin, RejectError),
+        pbs.EXECJOB_END: (handle_execjob_end, OfflineError)
+    }
+
+    handler, timeout_exc = handlers.get(event.type, (None, None))
+    if not handler:
+        log_with_caller(pbs.EVENT_ERROR, '%s event is not handled by this hook'
+                        % event.type)
+        event.accept()
+    try:
+        handler()
+    except KeyboardInterrupt:
+        raise timeout_exc
+
+
+if __name__ == '__builtin__':
+    try:
+        main()
+    except OfflineError:
+        vnode = HookHelper.get_event().vnode_list[pbs.get_local_nodename()]
+        vnode.state = pbs.ND_OFFLINE
+        vnode.comment = 'Node offlined due to Health Check Failure'
```

Comment:
Instead of Health Check Failure, maybe we should say "Node offlined due to timeout of PBS_cray_jacs hook"?
Or if we can, maybe even tell them which timeout (post or delete)?

Reply:
According to the design, if the POST encounters a timeout, the hook rejects, it does not offline. Should I change the design?
I've added the exception message to the vnode comment.

Reply:
oh, that's right, I forgot that POST timeouts/failures doesn't offline nodes.